Common/DataMonitor.cs

```diff
@@ -0,0 +1,265 @@
+            var isUniverseData = path.Contains("coarse", StringComparison.OrdinalIgnoreCase) || 
+                path.Contains("universe", StringComparison.OrdinalIgnoreCase);
+
+            if (e.Succeded)
+            {
+                WriteLineToFile(_succeededDataRequestsWriter, path, _succeededDataRequestsFileName);
+                Interlocked.Increment(ref _succeededDataRequestsCount);
+                if (isUniverseData)
+                {
+                    Interlocked.Increment(ref _succeededUniverseDataRequestsCount);
+                }
+            }
+            else
+            {
+                WriteLineToFile(_failedDataRequestsWriter, path, _failedDataRequestsFileName);
+                Interlocked.Increment(ref _failedDataRequestsCount);
+                if (isUniverseData)
+                {
+                    Interlocked.Increment(ref _failedUniverseDataRequestsCount);
+                }
+
+                if (Logging.Log.DebuggingEnabled)
+                {
+                    Logging.Log.Debug($"DataMonitor.OnNewDataRequest(): Data from {path} could not be fetched");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Terminates the data monitor generating a final report
+        /// </summary>
+        public void Exit()
+        {
+            if (_exited)
+            {
+                return;
+            }
+
+            _requestRateCalculationThread.StopSafely(TimeSpan.FromSeconds(5), _cancellationTokenSource);
+            _succeededDataRequestsWriter.Close();
+            _failedDataRequestsWriter.Close();
+            _exited = true;
+
+            StoreDataMonitorReport(GenerateReport());
+            
+            _succeededDataRequestsCount = 0;
+            _failedDataRequestsCount = 0;
+            _requestRates.Clear();
+            _prevRequestsCount = 0;
+            _lastRequestRateCalculationTime = default;
+        }
+
+        public void Dispose()
+        {
+            _succeededDataRequestsWriter.Close();
+            _succeededDataRequestsWriter.DisposeSafely();
+            _failedDataRequestsWriter.Close();
+            _failedDataRequestsWriter.DisposeSafely();
+            _cancellationTokenSource?.DisposeSafely();
+        }
+
+        protected virtual string StripDataFolder(string path)
+        {
+            if (path.StartsWith(Globals.DataFolder, StringComparison.OrdinalIgnoreCase))
+            {
+                return path.Substring(Globals.DataFolder.Length);
+            }
+
+            return path;
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="DataMonitor"/> instance
+        /// </summary>
+        private void Initialize()
+        {
+            if (_requestRateCalculationThread != null)
+            {
+                return;
+            }
+            
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            _requestRateCalculationThread = new Thread(() =>
```

Comment:
Cool! we should take a lock here to make sure we start a single one, after the initial null check, there could be multiple threads fetching files at the same time 👍